hudi-common/src/main/java/org/apache/hudi/metadata/HoodieBackedTableMetadata.java

```diff
@@ -636,6 +616,34 @@ private <T> ClosableIterator<T> lookupRecords(List<String> sortedKeys,
     }
   }
 
+  private Predicate buildPrefixPredicate(String partitionName, List<String> sortedKeys) {
+    if (MetadataPartitionType.fromPartitionPath(partitionName)
+          .equals(MetadataPartitionType.SECONDARY_INDEX)) {
+      return Predicates.startsWithAny(null,
+          sortedKeys.stream()
+          .map(SecondaryIndexKeyUtils::constructSecondaryIndexKeyPrefix)
+          .map(Literal::from)
+          .collect(Collectors.toList()));
+    }
+    return Predicates.startsWithAny(null, sortedKeys.stream()
+        .map(Literal::from)
+        .collect(Collectors.toList()));
+  }
+
+  private Predicate  buildPredicate(String partitionName, List<String> sortedKeys) {
+    if (MetadataPartitionType.fromPartitionPath(partitionName)
+          .equals(MetadataPartitionType.SECONDARY_INDEX)) {
+      return Predicates.startsWithAny(null,
+          sortedKeys.stream()
+          .map(SecondaryIndexKeyUtils::constructSecondaryIndexKeyPrefix)
+          .map(Literal::from)
+          .collect(Collectors.toList()));
+    }
```

Comment:
This method is for full-key lookup, so secondary index lookup with prefixes should not show up here, i.e., there is no need for the `if` condition.

It would be better to have one `buildPredicate` method to consider both full and prefix key lookup.  The current set of changes are a mixed approach.